Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1415,78 +1505,57 @@ class DecoratorTransform(ScopeTrackingTransform, SkipDeclarations):
     with up to the three getter, setter and deleter DefNodes.
     The functional style isn't supported yet.
     """
-    _properties = None
+    _properties_node_sets = None
 
-    _map_property_attribute = {
+    _property_attributes = {
         'getter': EncodedString('__get__'),
         'setter': EncodedString('__set__'),
         'deleter': EncodedString('__del__'),
-    }.get
+    }
+
+    _map_property_attribute = _property_attributes.get
 
     def visit_CClassDefNode(self, node):
-        if self._properties is None:
-            self._properties = []
-        self._properties.append({})
+        if self._properties_node_sets is None:
+            self._properties_node_sets = []
+        id_props_transform = _IdentifyPropertiesTransform(self.context,
+                                                            self.scope_type,
+                                                            self.scope_node,
+                                                            self.current_directives)
+        id_props_transform(node)
+
+        # transform @property decorators
+        properties = id_props_transform.properties
+        self._properties_node_sets.append(set())
+        for name, property_dict in properties.items():
+            self._add_property(node, name, property_dict)
+
         node = super(DecoratorTransform, self).visit_CClassDefNode(node)
-        self._properties.pop()
+        self._properties_node_sets.pop()
         return node
 
     def visit_PropertyNode(self, node):
         # Low-level warning for other code until we can convert all our uses over.
-        level = 2 if isinstance(node.pos[0], str) else 0
-        warning(node.pos, "'property %s:' syntax is deprecated, use '@property'" % node.name, level)
+        if not node.created_from_decorator:
+            level = 2 if isinstance(node.pos[0], str) else 0
+            warning(node.pos, "'property %s:' syntax is deprecated, use '@property'" % node.name, level)
```

Comment:
I wonder if the parser wouldn't be a better place to emit the warning. It knows for sure what syntax was used.

Reply:
Done - that lets me get rid of `created_from_decorator` too.

I'm not sure `isinstance(node.pos[0], str)` was working right anyway - it's not a `FileSourceDescriptor`. So we may not have been showing the warning to anyone for a while.

Reply:
> I'm not sure `isinstance(node.pos[0], str)` was working right anyway - it's not a `FileSourceDescriptor`. So we may not have been showing the warning to anyone for a while.

This might have been meant to exclude internal utility code, for which the source is a string, something like `"<string>"` or so.

Reply:
Yes - it definitely was meant to exclude internal utility code. It now looks like a choice between `StringSourceDescriptor` (for the internal code) and `FileSourceDescriptor` for user code. But I'm not sure what it would have been in the past.